File: src/main/scala/org/anz/codechallenge/validators/RecordCountValidator.scala
package org.anz.codechallenge.validators

import com.google.gson.Gson
import org.anz.codechallenge.schema.{Schema, Metadata, Tag}
import org.apache.spark.sql.DataFrame

import scala.io.Source

/**
  * Record count validator
  * @param inputMetadata - Input metadata representing metadata info of input file
  * @param file_schema - File schema representing schema of the input file
  * @param tagFile - Tagfile representing he file name and record count
  * @param dataFrame - Input file read as dataframe which needs validation
  */
class RecordCountValidator (
                             inputMetadata: Metadata,
                             file_schema: Schema,
                             tagFile: Tag,
                             dataFrame: DataFrame
                           ) extends Validator {

  /**
    * Perform record count validation
    * @return - status of validation
    */
  override def validate(): String = {
    println("Performing recordcount validation")

    val actualCount = dataFrame.count()
    val expectedCount = tagFile.record_count

    println("input data records count "+actualCount)
    println("expected records count "+expectedCount)

    val status = if(actualCount == expectedCount)  "0" else "1"
    status

    println("Record count validation status is "+status)
    status
  }

}

File: src/main/java/org/anz/codechallenge/validators/RecordCountValidator.java
package org.anz.codechallenge.validators;

import org.anz.codechallenge.filedetails.ContentParams;
import org.anz.codechallenge.filedetails.FileContent;
import org.anz.codechallenge.schema.JSONSchema;
import org.anz.codechallenge.tags.DelimitedTag;
import org.anz.codechallenge.tags.Tag;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;

public class RecordCountValidator implements Validator {

    private FileContent fileContent;

    public RecordCountValidator(FileContent fileContent) {
        this.fileContent = fileContent;
    }

    @Override
    public String validate() {
        System.out.println("Performing recordcount validation");

        Tag tagFile = fileContent.getFileMetadata().getTagFile();
        String status = "1";

        if(tagFile instanceof DelimitedTag) {
            long actualCount = fileContent.getDataframe().count();
            int expectedCount = ((DelimitedTag) fileContent.getFileMetadata().getTagFile()).getRecord_count();

            System.out.println("input data records count " + actualCount);
            System.out.println("expected records count " + expectedCount);

            status = (actualCount == expectedCount) ? "0" : "1";
        }
        System.out.println("Record count validation status is "+status);
        return status;
    }
}
